pico_w/wifi/freertos/ntp_client_socket/picow_freertos_ntp_client_socket.c

```diff
@@ -0,0 +1,174 @@
+/**
+ * Copyright (c) 2022 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "pico/cyw43_arch.h"
+#include "pico/stdlib.h"
+
+#include "lwip/ip4_addr.h"
+#include "lwip/dns.h"
+#include "lwip/pbuf.h"
+#include "lwip/udp.h"
+#include "lwip/sockets.h"
+#include "lwip/netdb.h"
+
+#include "FreeRTOS.h"
+#include "task.h"
+
+#ifndef RUN_FREERTOS_ON_CORE
+#define RUN_FREERTOS_ON_CORE 0
+#endif
+
+#define TEST_TASK_PRIORITY                ( tskIDLE_PRIORITY + 1UL )
+
+#define NTP_SERVER ("pool.ntp.org")
+#define NTP_MSG_LEN (48)
+#define NTP_PORT (123)
+// seconds between 1 Jan 1900 and 1 Jan 1970
+#define NTP_DELTA (2208988800)
+#define ntpEST_TIME (30 * 1000)
+#define NTP_FAIL_TIME (10)
+
+void main_task(__unused void *params) {
+    if (cyw43_arch_init()) {
+        printf("failed to initialise\n");
+        vTaskDelete(NULL);
+    }
+    cyw43_arch_enable_sta_mode();
+    printf("Connecting to Wi-Fi...\n");
+    if (cyw43_arch_wifi_connect_timeout_ms(WIFI_SSID, WIFI_PASSWORD, CYW43_AUTH_WPA2_AES_PSK, 30000)) {
+        printf("failed to connect.\n");
+        vTaskDelete(NULL);
+    } else {
+        printf("Connected.\n");
+    }
+
+    while(true) {
+        const struct addrinfo ntp_info = {
+            .ai_flags = 0,
+            .ai_family = AF_UNSPEC,
+            .ai_socktype = SOCK_DGRAM,
+            .ai_protocol = 0,
+        };
+
+        struct addrinfo *dns_result = NULL;
+        printf("Trying DNS...\n");
+        int err = getaddrinfo(NTP_SERVER, "123", &ntp_info, &dns_result);
+        if (err != 0)
+        {
+            printf("DNS lookup failed with error: %d\n", err);
+            continue;
+        }
+#ifndef INET6_ADDRSTRLEN
+        char address[INET_ADDRSTRLEN] = {0};
+#else
+        char address[INET6_ADDRSTRLEN] = {0};
+#endif
+        inet_ntop(dns_result->ai_family, (struct sockaddr_in*)(dns_result->ai_addr), address, sizeof(address));
```

Comment:
I don't think this is correct, inet_ntop is expecting an ip_addr_t not a sockaddr. You have to call inet_addr_to_ip4addr or inet6_addr_to_ip6addr